### PR TITLE
CODAP-1357: allow cross-attribute prev() references without cycle error

### PR DESCRIPTION
**Pull request by bfinzer**
## Summary

Fixes [CODAP-1357](https://concord-consortium.atlassian.net/browse/CODAP-1357).

Two attributes that reference each other only through `prev()` (e.g. the Markov-style
sunny/rainy formulas in the linked story) were incorrectly flagged as a circular
reference. `prev()` reads the previous case's value, so there's no actual row-level
cycle - V2 handles this without error.

The dependency extractor (`formula-dependency-utils.ts`) already filtered out direct
self-references inside `prev()`, but cross-references between two attributes still
landed in the dependency graph as a mutual cycle. This change marks such dependencies
with `selfReferenceAllowed: true` on the `ILocalAttributeDependency` record and skips
them in the cycle detector, while still keeping them in the dependency list so
observers fire when the referenced attribute changes. The flag is cleared if the same
attribute is also referenced outside a `prev()` call elsewhere in the formula, so
genuine cycles paired with a `prev()` reference are still caught.

A follow-up issue (CODAP-1358) was filed for extending the same fix to `next()`,
which has the same row-shifted semantics but a different runtime path that needs
separate verification.

## Test plan

- [x] New unit tests in `attribute-formula-adapter.test.ts` cover the CODAP-1357 sunny/rainy case and an edge case where a `prev()` cross-reference is paired with a direct reference (real cycle should still be detected).
- [x] All 462 existing formula unit tests pass.
- [x] Lint and TypeScript check pass.
- [ ] Manually verified the attached `markovWeatherTest.codap` document now opens without circular-reference errors in V3.
- [ ] Cypress regression tests pass on CI (will trigger via `run regression` label once draft CI is green).

[CODAP-1357]: https://concord-consortium.atlassian.net/browse/CODAP-1357?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: v3/src/models/formula/utils/formula-dependency-utils.ts [hunk 1]
```diff
@@ -45,15 +45,30 @@ export const getFormulaDependencies = (formulaCanonical: string, formulaAttribut
...
-                isLocalAttributeDependency(dependency) &&
-                dep.attrId === dependency.attrId
-      })
-      if (dependency && (!isSelfReference || !isSelfReferenceAllowed) && !isPresent) {
-        result.push(dependency)
+      // Direct self-reference inside prev() (e.g. CumulativeValue = Value + prev(CumulativeValue, 0))
+      // is dropped from the dependency list - prev() breaks the row-level cycle, and adding the
+      // attribute would create an observer cycle that recalculates the formula from its own output.
+      const skipDirectSelfRef = isSelfReference && isSelfReferenceAllowed
+      if (dependency && !skipDirectSelfRef) {
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/models/formula/utils/formula-dependency-utils.ts:67]: Good catch - this was a real corner-case regression. Fixed by aligning `isSelfReferenceAllowed` propagation with `isSemiAggregate`: only the row-shifted args (expression and filter, per `prev`'s `[true, false, true]`) get marked, not `defaultValue`. Added a regression test (`prev(B, B)` paired with `B = A`) that fails without the fix.

File: v3/src/models/formula/utils/formula-dependency-utils.ts [hunk 2]
```diff
@@ -45,15 +45,30 @@ export const getFormulaDependencies = (formulaCanonical: string, formulaAttribut
...
+          ? result.find((dep): dep is ILocalAttributeDependency =>
+              isLocalAttributeDependency(dep) && dep.attrId === dependency.attrId)
+          : undefined
+        if (existing) {
+          // If the same attribute is referenced outside a selfReferenceAllowed function anywhere
+          // in the formula, the dependency cannot be treated as cycle-safe.
+          if (!isSelfReferenceAllowed) {
+            existing.selfReferenceAllowed = false
+          }
+        } else {
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/models/formula/utils/formula-dependency-utils.ts:82]: Good catch - although this was pre-existing behavior, fixing it here is the right call since we're touching the dedup logic. Now merging `aggregate` as an OR when an existing entry is found.

File: v3/src/models/formula/attribute-formula-adapter.test.ts [hunk 3]
```diff
@@ -123,6 +123,37 @@ describe("AttributeFormulaAdapter", () => {
...
       expect(error).toBeUndefined()
     })
+
+    it("should not report false cycle when two attributes reference each other only via prev() (CODAP-1357)", () => {
+      // Markov-style mutual recurrence: each attribute references the other only through prev().
+      // prev() reads the previous case, so this is not a runtime cycle.
+      const { dataSet, adapter, contextMap, metadataMap } = getCycleTestEnv([
+        { name: "sunny", formula: "caseIndex=1 ? 3600 : 5/6 * prev(sunny) + 4/6 * prev(rainy)" },
+        { name: "rainy", formula: "caseIndex=1 ? 0 : 1/6 * prev(sunny) + 2/6 * prev(rainy)" }
+      ])
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/models/formula/attribute-formula-adapter.test.ts:133]: Considered, but leaving as-is. The dependency extractor only inspects symbol nodes; operator form doesn't affect what's tested here (and MathJS parses `=`/`?:` cleanly enough that the existing cycle-detection tests in this file have used this same shorthand). Worth a follow-up if we want to broaden the test family.

File: v3/src/models/formula/utils/formula-dependency-utils.ts [hunk 4]
```diff
@@ -45,15 +60,34 @@ export const getFormulaDependencies = (formulaCanonical: string, formulaAttribut
...
+          if (isLocalAttributeDependency(dependency) && isSelfReferenceAllowed) {
+            // Cross-attribute reference inside prev() (e.g. sunny = prev(rainy), rainy = prev(sunny)).
+            // Keep the dependency so observers fire when the referenced attribute changes, but mark
+            // it so the cycle detector ignores it - prev() breaks the row-level cycle.
+            dependency.selfReferenceAllowed = true
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/models/formula/utils/formula-dependency-utils.ts:87]: Good catch - this was a genuine runtime regression. With cycle detection relaxed for mutual `prev()` deps, those formulas now reach `recalculateDownstreamFormulas` (previously blocked at validation), and there's no visited guard, so the cascade recurses indefinitely. Fixed by adding a re-entry guard on `FormulaManager.recalculateFormula` that skips a formula already mid-recalculation in the current call stack. Added a regression test in `formula-manager.test.ts` that reproduces the stack overflow without the guard (verified by temporarily reverting the guard - the test correctly fails with the expected RangeError trace).

**Review by kswenson — APPROVED** [review]
👍 Looks good -- the following review was developed in conjunction with Claude Code 🤖.

## Review Summary

**What it does:** Fixes the false-positive cycle error for mutual `prev()` references (e.g. Markov sunny/rainy) while still catching real cycles. `prev()` reads the previous case's value, so the row-level cycle is broken; V2 has always handled this without error.

**The original approach** (commit 8c45244):
1. New `selfReferenceAllowed?: boolean` on `ILocalAttributeDependency`.
2. Dependency extractor marks cross-attribute deps inside `prev()` and clears the flag if the same attribute is also referenced outside `prev()` elsewhere.
3. Cycle detector filters flagged deps; downstream cascade preserved for observers.

**Refinements added during review:**
- **ad16f91e:** `selfReferenceAllowed` now propagates only to row-shifted args via `isSemiAggregate` — `prev()`'s `defaultValue` (current-case context) is no longer treated as cycle-safe. Also merges the `aggregate` flag on dedup. New regression test for `A = prev(B, B), B = A`.
- **3e0e861f:** Re-entry guard on `FormulaManager.recalculateFormula` to prevent infinite recursion in the downstream cascade through cycle-safe edges. Previously, mutual `prev()` deps would have stack-overflowed `recalculateDownstreamFormulas` at runtime (the pre-PR cycle detector had prevented this code path from executing). New regression test that reproduces the overflow without the guard.

464 formula tests pass, TS clean, lint clean, full Cypress regression green.
